community/index/src/main/java/org/neo4j/index/gbptree/GBPTree.java

```diff
@@ -465,19 +465,36 @@ public long acquireNewId() throws IOException
     @Override
     public void checkpoint( IOLimiter ioLimiter ) throws IOException
     {
+        // Flush dirty pages of the tree, do this before acquiring the lock so that writers won't be
+        // blocked while we do this
         pagedFile.flushAndForce( ioLimiter );
+
+        // Block writers, or if there's a current writer then wait for it to complete and then block
+        // From this point and till the lock is released we know that the tree won't change.
         writerCheckpointMutex.lock();
         try
         {
+            // Flush dirty pages since that last flush above. This should be a very small set of pages
+            // and should be rather fast. In here writers are blocked and we want to minimize this
+            // windows of time as much as possible, that's why there's an initial flush outside this lock.
             pagedFile.flushAndForce( ioLimiter );
```

Comment:
To minimise the amount of time in this critical section, use `IOLimiter.unlimited()`.